### ISSUE TITLE
ECFTrustManager fails to verify a certificate chain if it is out-of-order

### ISSUE THREAD

**Issue by mdaloia** [eclipse/ecf#27]
The [`ECFTrustManager.verify()`](https://github.com/eclipse/ecf/blob/696f197b4131f84c416135cd6117384567b0da56/framework/bundles/org.eclipse.ecf.ssl/src/org/eclipse/ecf/internal/ssl/ECFTrustManager.java#L53-L74) method expects that the certificates are presented in a specific order by the server.

Sometimes there are servers/proxies that present them out-of-order.
Browsers, OpenSSL and maybe other tools are less restrictive and validates them even if they are out-of-order.

The current implementation it is really compliant with the spec of TLSv1.2 ([RFC 5246, section 7.4.2](https://www.rfc-editor.org/rfc/rfc5246#section-7.4.2))

```
   certificate_list
      This is a sequence (chain) of certificates.  The sender's
      certificate MUST come first in the list.  Each following
      certificate MUST directly certify the one preceding it.  Because
      certificate validation requires that root keys be distributed
      independently, the self-signed certificate that specifies the root
      certificate authority MAY be omitted from the chain, under the
      assumption that the remote end must already possess it in order to
      validate it in any case.
```

However, with the TLSv1.3 ([RFC 8446, section 4.4.2](https://www.rfc-editor.org/rfc/rfc8446#section-4.4.2)) this restriction have been removed:

```
   certificate_list:  A sequence (chain) of CertificateEntry structures,
      each containing a single certificate and set of extensions.

   Note: Prior to TLS 1.3, "certificate_list" ordering required each
   certificate to certify the one immediately preceding it; however,
   some implementations allowed some flexibility.  Servers sometimes
   send both a current and deprecated intermediate for transitional
   purposes, and others are simply configured incorrectly, but these
   cases can nonetheless be validated properly.  For maximum
   compatibility, all implementations SHOULD be prepared to handle
   potentially extraneous certificates and arbitrary orderings from any
   TLS version, with the exception of the end-entity certificate which
   MUST be first.
```

Source: https://systemrequest.net/index.php/304/?show=305#a305

I've made a local test with this configuration:

- Tried nginx and Apache HTTPD Docker images
- Configured to support only TLSv1.2
- Configured with a certificate (and its intermediate ones) and its private key
	- Used a certificate chain with 2 intermediates presented in switched order:
	  ```
	  subject: end-user.foo.bar
	  issuer: int2.foo.bar

	  subject: int1.foo.bar
	  issuer: ca.foo.bar

	  subject: int2.foo.bar
	  issuer: int1.foo.bar

	  subject: ca.foo.bar
	  issuer: ca.foo.bar
	  ```
- Added the root certificate into the `cacerts` file of the JVM running Eclipse
- Configured to serve a p2 repository
- Added the custom domain (`end-user.foo.bar` in this example) to the hosts file (`/etc/hosts`)

If from a browser I access to https://end-user.foo.bar:8443/ it serves a simple `index.html` file without any issue or certificate warning.

But when I add the URL of this p2 repository in the `Available Software Sites` and hit the `Reload` button I get this exception:

```
org.eclipse.equinox.p2.core.ProvisionException: Unable to read repository at https://end-user.foo.bar:8443/artifacts.xml.
	at org.eclipse.equinox.internal.p2.repository.CacheManager.createCache(CacheManager.java:249)
	at org.eclipse.equinox.internal.p2.artifact.repository.simple.SimpleArtifactRepositoryFactory.getLocalFile(SimpleArtifactRepositoryFactory.java:68)
	at org.eclipse.equinox.internal.p2.artifact.repository.simple.SimpleArtifactRepositoryFactory.load(SimpleArtifactRepositoryFactory.java:92)
	at org.eclipse.equinox.internal.p2.artifact.repository.simple.SimpleArtifactRepositoryFactory.load(SimpleArtifactRepositoryFactory.java:80)
	at org.eclipse.equinox.internal.p2.artifact.repository.ArtifactRepositoryManager.factoryLoad(ArtifactRepositoryManager.java:92)
	at org.eclipse.equinox.internal.p2.repository.helpers.AbstractRepositoryManager.loadRepository(AbstractRepositoryManager.java:787)
	at org.eclipse.equinox.internal.p2.repository.helpers.AbstractRepositoryManager.loadRepository(AbstractRepositoryManager.java:685)
	at org.eclipse.equinox.internal.p2.repository.helpers.AbstractRepositoryManager.basicRefreshRepository(AbstractRepositoryManager.java:205)
	at org.eclipse.equinox.internal.p2.artifact.repository.ArtifactRepositoryManager.refreshRepository(ArtifactRepositoryManager.java:136)
	at org.eclipse.equinox.p2.ui.RepositoryManipulationPage.lambda$17(RepositoryManipulationPage.java:596)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:122)
Caused by: javax.net.ssl.SSLHandshakeException: Certificate chain is not valid
	at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:131)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:353)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:296)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:291)
	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(CertificateMessage.java:654)
	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.onCertificate(CertificateMessage.java:473)
	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.consume(CertificateMessage.java:369)
	at java.base/sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:392)
	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:443)
	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:421)
	at java.base/sun.security.ssl.TransportContext.dispatch(TransportContext.java:183)
	at java.base/sun.security.ssl.SSLTransport.decode(SSLTransport.java:172)
	at java.base/sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1506)
	at java.base/sun.security.ssl.SSLSocketImpl.readHandshakeRecord(SSLSocketImpl.java:1416)
	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:456)
	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:427)
	at org.apache.hc.client5.http.ssl.SSLConnectionSocketFactory.createLayeredSocket(SSLConnectionSocketFactory.java:273)
	at org.apache.hc.client5.http.ssl.SSLConnectionSocketFactory.connectSocket(SSLConnectionSocketFactory.java:241)
	at org.apache.hc.client5.http.impl.io.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:148)
	at org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:396)
	at org.apache.hc.client5.http.impl.classic.InternalExecRuntime.connectEndpoint(InternalExecRuntime.java:158)
	at org.apache.hc.client5.http.impl.classic.InternalExecRuntime.connectEndpoint(InternalExecRuntime.java:168)
	at org.apache.hc.client5.http.impl.classic.ConnectExec.execute(ConnectExec.java:136)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(ExecChainElement.java:51)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement$1.proceed(ExecChainElement.java:57)
	at org.apache.hc.client5.http.impl.classic.ProtocolExec.execute(ProtocolExec.java:175)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(ExecChainElement.java:51)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement$1.proceed(ExecChainElement.java:57)
	at org.apache.hc.client5.http.impl.classic.HttpRequestRetryExec.execute(HttpRequestRetryExec.java:96)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(ExecChainElement.java:51)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement$1.proceed(ExecChainElement.java:57)
	at org.apache.hc.client5.http.impl.classic.ContentCompressionExec.execute(ContentCompressionExec.java:133)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(ExecChainElement.java:51)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement$1.proceed(ExecChainElement.java:57)
	at org.apache.hc.client5.http.impl.classic.RedirectExec.execute(RedirectExec.java:115)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(ExecChainElement.java:51)
	at org.apache.hc.client5.http.impl.classic.InternalHttpClient.doExecute(InternalHttpClient.java:170)
	at org.apache.hc.client5.http.impl.classic.CloseableHttpClient.execute(CloseableHttpClient.java:75)
	at org.eclipse.ecf.provider.filetransfer.httpclient5.HttpClientFileSystemBrowser.runRequest(HttpClientFileSystemBrowser.java:251)
	at org.eclipse.ecf.provider.filetransfer.browse.AbstractFileSystemBrowser$DirectoryJob.run(AbstractFileSystemBrowser.java:71)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: org.eclipse.ecf.internal.ssl.ECFCertificateException: Certificate chain is not valid
	at org.eclipse.ecf.internal.ssl.ECFTrustManager.verify(ECFTrustManager.java:68)
	at org.eclipse.ecf.internal.ssl.ECFTrustManager.checkServerTrusted(ECFTrustManager.java:32)
	at java.base/sun.security.ssl.AbstractTrustManagerWrapper.checkServerTrusted(SSLContextImpl.java:1510)
	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(CertificateMessage.java:638)
	... 36 more
Caused by: java.security.SignatureException: Signature does not match.
	at java.base/sun.security.x509.X509CertImpl.verify(X509CertImpl.java:422)
	at java.base/sun.security.x509.X509CertImpl.verify(X509CertImpl.java:390)
	at org.eclipse.ecf.internal.ssl.ECFTrustManager.verify(ECFTrustManager.java:65)
	... 39 more
```

Even if we still don't have TLSv1.3 support (see #26) we can make it a little less restrictive as other tools out there.

If there are some concern, then maybe let enable it explicitly with a System Property and use the current implementation by default. Keep in mind that eventually this will be need for the TLSv1.3 support by default.

**Comment by scottslewis** [comment]
Thanks for the info.

I'm not clear on what the fix is.  Could you describe the fix?  (to ECFTrustManager?)

Thanks.

**Comment by mdaloia** [comment]
Thanks @scottslewis for your time reviewing it! Really appreciated!

Yes, I think that ECFTrustManager should be enhanced.

I guess that, before calling the `verify()`, the `X509Certificate[]` could be pre-processed to reorder the certificates and discard the irrelevant ones (if they are present).

Based on the RFC, "_the end-entity certificate which MUST be first_.", that is a hard rule that all servers must comply when they send the certificates... Using the Issuer value of that end-entity certificate we need to find a certificate that matches with its Subject. Next we should take the found certificate and repeat the same procedure of matching the Issuer of this certificate with the Subject of another until no certificate is found or the root certificate is found (the one that Issuer and Subject are the same... It may or may not be in the chain).
In the end we will get the certificate chain to validate ordered and without irrelevant certs (if there are one).

Maybe I'm oversimplifying it and missing something but I think that should be something similar to that.

If we take the previous example where certs 2 and 3 are inverted:

```
subject: end-user.foo.bar
issuer: int2.foo.bar

subject: int1.foo.bar
issuer: ca.foo.bar

subject: int2.foo.bar
issuer: int1.foo.bar

subject: ca.foo.bar
issuer: ca.foo.bar
```

or also where they are very unordered and with irrelevant certs:

```
subject: end-user.foo.bar
issuer: int2.foo.bar

subject: ca.foo.bar
issuer: ca.foo.bar

subject: int1.foo.bar
issuer: ca.foo.bar

subject: not-relevant-1.foo.bar
issuer: int1.foo.bar

subject: not-relevant-2.foo.bar
issuer: not-relevant-3.foo.bar

subject: int2.foo.bar
issuer: int1.foo.bar
```

the logic should order and filter it to:
```
subject: end-user.foo.bar
issuer: int2.foo.bar

subject: int2.foo.bar
issuer: int1.foo.bar

subject: int1.foo.bar
issuer: ca.foo.bar

subject: ca.foo.bar
issuer: ca.foo.bar
```

Is this reasoning logic?
Do you have other things in mind?

Let me know! Thanks!

**Comment by scottslewis** [comment]
Hi @mdaloia 

Just FYI:   The ECFTrustManager itself was written quite a few years ago...and not by me.  It may very well be the case that the standards/RFCs have changed or been enhanced since it was written, and we/I just haven't kept up with the standards changes.  So your points about needed enhancements are well-taken.

I don't see anything wrong with the logic you describe above wrt enhancing ECFTrustManager to modify the cert search order.  In looking at the existing code in ECFTrustManager.checkServerTrusted (which is currently the only API caller of the private verify method) it seems to me that you could introduce a new private method, called just prior to verify, which makes the changes you describe to the cert order (and filtering).

Would you be able to work with the code to produce a pull request?  If so you should probably fork the entire ECF repo to a your own account, and then create a new branch on your fort (perhaps with name including issue-27...i.e. this issue).   Then you can test the fix in your environment from within Eclipse.  

ECF currently does not have any test code that tests the ECFTrustManager or the org.eclipse.ecf.ssl bundle that contains that class...but if you end up creating some in your environment I can/will add a junit test it to the ECF unit tests (which are in the ecf/tests part of repo).

wdyt?

**Comment by scottslewis** [comment]
Hi @mdaloia .   I've merged pr #39 .  Thanks much for the contribution.    Is there more to do on this?   ECF is going to have to contribute 3.14.38 to Eclipse fairly soon and I'm expecting everything as it is will be part of 3.14.38...but is there more to do here before making it non-experimental and/or considering this issue closed?   Looking to you for guidance about that.   Of course ECF can make contributions in the future, but they will probably have to wait until after the June simultaneous release.

**Comment by mdaloia** [comment]
Hi @scottslewis, I currently don't have any other plans to change the implementation that I did.
I think it can be incorporated into the next release because it is disabled by default so it is a safe change.

Let me know if anything else needed please let me know.

Thanks for the follow-up!
